book/02-variables-and-functions.md

```diff
@@ -0,0 +1,886 @@
+
+ But this decision has some good effects. For one thing, recursive (and
+especially mutually recursive) definitions are harder to reason about than
+nonrecursive ones. It's therefore useful that, in the absence of an explicit
+`rec`, you can assume that a `let` binding is nonrecursive, and so can only
+build upon previous definitions.
+
+ In addition, having a nonrecursive form makes it easier to create a new
+definition that extends and supersedes an existing one by shadowing it.
+
+#### Prefix and Infix Operators {#prefix-and-infix-operators data-type="sect2"}
+
+So far, we've seen examples of functions used in both prefix and infix
+style.<idx>operators/prefix and infix operators</idx><idx>infix
+operators</idx><idx>prefix operators</idx><idx id="FNCprf">functions/prefix
+and infix operators</idx>
+
+<link rel="import" href="code/variables-and-functions/main.mlt" part=
+"24" />
+
+You might not have thought of the second example as an ordinary function, but
+it very much is. Infix operators like `+` really only differ syntactically
+from other functions. In fact, if we put parentheses around an infix
+operator, you can use it as an ordinary prefix function.
+
+<link rel="import" href="code/variables-and-functions/main.mlt" part=
+"25" />
+
+In the second expression, we've partially applied `(+)` to create a function
+that increments its single argument by `3`.
+
+A function is treated syntactically as an operator if the name of that
+function is chosen from one of a specialized set of identifiers. This set
+includes identifiers that are sequences of characters from the following set:
+
+<link rel="import" href="code/variables-and-functions/operators.syntax" />
+
+or is one of a handful of predetermined strings, including `mod`, the modulus
+operator, and `lsl`, for "logical shift left," a bit-shifting operation.
+
+We can define (or redefine) the meaning of an operator. Here's an example of
+a simple vector-addition operator on `int` pairs.
+
+<link rel="import" href="code/variables-and-functions/main.mlt" part=
+"26" />
+
+Note that you have to be careful when dealing with operators containing 
+`*`. Consider the following example.
+
+<link rel="import" href="code/variables-and-functions/main.mlt" part=
+"27" />
+
+What's going on is that `(***)` isn't interpreted as an operator at all; it's
+read as a comment! To get this to work properly, we need to put spaces around
+any operator that begins or ends with `*`.
+
+<link rel="import" href="code/variables-and-functions/main.mlt" part=
+"28" />
+
+The syntactic role of an operator is typically determined by its first
+character or two, though there are a few exceptions.
+[Table2_1](02-variables-and-functions.html#table2_1){data-type=xref} breaks
+the different operators and other syntactic forms into groups from highest to
+lowest precedence, explaining how each behaves syntactically. We write 
+`!`... to indicate the class of operators beginning with `!`.
+
```

Comment:
Nothing to do, but it's worth noting that these tables are pretty hard to read. Fixable with the auto-align functionality in emacs, I think.

Reply:
There are various ways to express tables in pandoc markdown: https://pandoc.org/MANUAL.html#tables

Happy to switch to something more readable (even if it is a bit of more work). We can also convert them manually directly.

Reply:
Agreed. The syntax is actually not  a problem from my perspective; it's just that it would be nice to have alignment. Agreed we can just add this by hand later.